handle masked_gather in late-gc-lowering

vchuravy opened the pull request:
fixes #34569

vchuravy committed:
[GCLowering] handle vectorized loads

Vectorized loads can come in 3 variants:
- scalar
- masked
- gather

All work about the same--if we run into a Loaded pointer, we need to
follow back the base pointer and track it. And otherwise they're just
normal def/use statements. But we also need to fix up after an LLVM
mistake and put a valid value in the pass-through slot.

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

vchuravy commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -828,6 +828,27 @@ std::vector<int> LateLowerGCFrame::NumberAllBase(State &S, Value *CurrentV) {
+    } else if (auto *II = dyn_cast<IntrinsicInst>(CurrentV)) {
+        switch (II->getIntrinsicID()) {
+            case Intrinsic::masked_gather: {
+                Numbers = NumberAll(S, II->getOperand(0));
```
I am simply forwarding the operand here, unsure if that is the right way of doing things, additionally should we handle
other masked intrinsics like `masked.scatter` and `masked.load`?

According to the Langref https://llvm.org/docs/LangRef.html#llvm-masked-gather-intrinsics there is also a passthrough
argument which would be a tracked pointer (or in this case it was undef).

vtjnash replied:
No, it’s equivalent to a load, so they’re both a normal generating instruction (just like a regular CallInst)

KristofferC commented:
Bump, can this be merged?

vtjnash committed:
[GCLowering] strip all alias information

Even non LoadInst / StoreInst (esp. vectorized loads and stores) can have this information
